src/cargo/util_schemas/core/package_id_spec.rs

```diff
@@ -98,6 +98,9 @@ impl PackageIdSpec {
             Some(version) => Some(version.parse::<PartialVersion>()?),
             None => None,
         };
+        if name.is_empty() {
+            bail!("package ID specification must have a name: `{spec}`");
+        }
         validate_package_name(name, "pkgid", "")?;
```

Comment:
Maybe we should put the empty check into this function?

Reply:
That seems like where it should belong but we already have that check be separate for manifests
```rust
    let package_name = package.name.trim();
    if package_name.is_empty() {
        bail!("package name cannot be an empty string")
    }

    validate_package_name(package_name, "package name", "")?;
```

This makes me think some more investigation is needed.


Note: empty feature names were disallowed as of #12928 but that was because we relied on the behavior of our TOML parser to disallow it which wasn't conformant behavior.

Reply:
I have checked all the references for this function and it seems okay to include the check in the validate_package_name. However, I have also discovered some bugs in those places. For example, if you attempt to type `cargo add @1.2.3`, you will receive a panic instead of an error. Therefore, I suggest we move forward with this pull request first. Once that's done, I can work on moving it into the validate_package_name and also add tests to cover these problematic cases. This PR is more focused on fixing bugs from package ID spec.

Reply:
As a heads up, I'm looking at touching the name validation code as well, so it'll be good to coordinate so we don't cause each other a lot of pain

Reply:
Got it. I will prioritize ensuring that empty names are not allowed. I will make sure I only change one thing each time and do not bring any off-topic changes. Hope that can help.